### PR TITLE
[ng] expose tb_server_data_source as interface

### PR DESCRIPTION
**Pull request by stephanwlee**
We can provide an alternative implementation of the DataSource by using
Angular DI but the provided implementations require some common APIs so
they do not get out of sync. By implementing an interface, we can
guarantee some correctness at the compilation time.

### CODE REVIEW

File: tensorboard/webapp/webapp_data_source/data_api_route_converter_test.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+import {TEST_ONLY, DataApiRouteConverter} from './data_api_route_converter';
+
+describe('data_api_route_converter test', () => {
+  describe('getUrl', () => {
+    // let hrefSpy = jasmine.Spy;
```
Comment (psybuzz) [inline tensorboard/webapp/webapp_data_source/data_api_route_converter_test.ts]: nit: drop

File: tensorboard/webapp/webapp_data_source/tb_server_data_source.ts [hunk 2]
```diff
@@ -14,29 +14,37 @@ limitations under the License.
...
 
+import {DataApiRouteConverter} from './data_api_route_converter';
+import {DataSource} from './data_source';
+
 /** @typehack */ import * as _typeHackRxjs from 'rxjs';
 
 @Injectable()
-export class TBServerDataSource {
+export class TBServerDataSource implements DataSource {
   // TODO(soergel): implements WebappDataSource
```
Comment (psybuzz) [inline tensorboard/webapp/webapp_data_source/tb_server_data_source.ts]: Can we now drop this TODO, or explain what more we plan to address later?

File: tensorboard/webapp/webapp_data_source/data_api_route_converter.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {Injectable} from '@angular/core';
+
+export interface ApiRouteContext {
```
Comment (psybuzz) [inline tensorboard/webapp/webapp_data_source/data_api_route_converter.ts]: Do we have plans for `ApiRouteContext` in OSS?  If not, is there some reason that TB variants must rely on this being in OSS?

File: tensorboard/webapp/webapp_data_source/tb_server_data_source.ts (hunk 2 above)
Comment (stephanwlee) [reply tensorboard/webapp/webapp_data_source/tb_server_data_source.ts]: Honestly... idk what it means and I rather not touch the part of the code that is not related. Feel free to file an issue :)

**Comment by stephanwlee** [comment]
You are correct. I decided to provide an alternative implementation of the tb_server_data_source instead. Thanks for the idea!

**Pull request closed without merging**